### PR TITLE
Improve the haddocks related to followers

### PR DESCRIPTION
**Pull request by dnadales**
Adds documentation to `Follower`, `IteratorBlockGCed`, and `ChainUpdate`. The added documentation is based on the consensus report.

Part of #4183

# Checklist

- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [x] The documentation has been properly updated
    - [x] New tests are added if needed and existing tests are updated
    - [x] Any changes affecting Consensus packages must have an entry in the appropriate `changelog.d` directory created using [`scriv`](https://github.com/input-output-hk/scriv). If in doubt, see the [Consensus release process](../ouroboros-consensus/docs/ReleaseProcess.md).
    - [x] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [x] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [x] Reviewer requested

### CODE REVIEW

File: ouroboros-network-api/src/Ouroboros/Network/Block.hs [hunk 1]
```diff
@@ -365,8 +365,21 @@ decodeTip decodeHeaderHash = do
...
+    -- | The 'RollBack' instruction indicates that the follower should perform a
+    -- rollback by first backtracking to a certain point.
+    --
+    -- If a follower should switch to a fork, then it will first receive a
+    -- 'RollBack' instruction followed by as many 'AddBlock' as necessary to
+    -- reach the tip of the new chain.
+    --
+    -- When the follower's (implicit) position is in the immutable part of the
+    -- chain, no rollback instructions will be encountered.
+  | RollBack (Point block)
```
[the remark this answers is not in the document]
Comment (dnadales) [reply ouroboros-network-api/src/Ouroboros/Network/Block.hs:382]: Oh, good point. Should I remove he reference to followers here, and move the followers specific info to `followerInstruction`?

[the remark this answers is not in the document]
Comment (amesgen) [reply ouroboros-network-api/src/Ouroboros/Network/Block.hs]: Yeah, that seems like a sensible place to move it to :+1:

[the remark this answers is not in the document]
Comment (dnadales) [reply ouroboros-network-api/src/Ouroboros/Network/Block.hs]: Done https://github.com/input-output-hk/ouroboros-network/pull/4372/commits/453820c211abf01b3aba8c786c66f55c6cf01f87

**Pull request merged**